File: Java/src/main/java/BaekJoon/BFS/HideAndSeek3_13549.java
package BaekJoon.BFS;

import java.io.FileInputStream;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

import static Util.Constants.INPUT;

public class HideAndSeek3_13549 {
    static int MAX = 100001;
    static boolean[] check;
    static int[] ax= {-1, 1, 2};
    public static void main(String[] args) throws Exception{
        System.setIn(new FileInputStream(INPUT));
        Scanner sc = new Scanner(System.in);
        final int N = sc.nextInt();
        final int K = sc.nextInt();
        check = new boolean[MAX + 1];
        Queue<pair> q = new LinkedList<>();
        check[N] = true;
        q.add(new pair(N, 0));
        while(!q.isEmpty()) {
            int cx = q.element().x;
            int cc = q.element().c;
            if(cx == K) {
                System.out.println(cc);
                return;
            }
            q.poll();
            for(int i = 0; i < 3; i++) {
                int nx = cx;
                if(i == 2) {
                    nx *= ax[i];
                    if(nx < 0 || nx > MAX - 1) continue;
                    if(!check[nx]) {
                        check[nx] = true;
                        q.add(new pair(nx, cc));
                    }
                } else {
                    nx += ax[i];
                    if(nx < 0 || nx > MAX - 1) continue;
                    if(!check[nx]) {
                        check[nx] = true;
                        q.add(new pair(nx, cc + 1));
                    }
                }
            }

        }
    }

    static class pair {
        public int x;
        public int c;
        pair(int x, int c) {
            this.x = x;
            this.c = c;
        }
    }
}

File: Kotlin/src/baekJoon/graph/bfs/HideAndSeek3_13549.kt
import java.io.FileInputStream
import java.util.*
import kotlin.math.min


class HideAndSeek3 {

    private lateinit var arr: Array<Int>
    private val MAX = 100001

    fun solution() {
        System.setIn(FileInputStream(INPUT))
        val br = System.`in`.bufferedReader()
        val (n, k) = br.readLine().split(" ").map { it.toInt() }
        arr = Array(MAX) { -1 }
        val q: Queue<Pair<Int, Int>> = LinkedList()
        q.add(n to 0)
        arr[n] = 0
        while(q.isNotEmpty()) {
            val cx = q.peek().first
            val cc = q.peek().second
            println("cx : $cx, cc : $cc")
            q.poll()
            if(cx == k) {
                println(min(arr[cx], cc))
                return
            }
            if(cx + 1 < MAX) {
                if(arr[cx + 1] != -1) {
                    if(arr[cx + 1] > cc + 1) {
                        arr[cx + 1] = cc + 1
                        q.add(cx + 1 to cc + 1)
                    }
                } else {
                    arr[cx + 1] = cc + 1
                    q.add(cx + 1 to cc + 1)
                }
            }
            if(cx - 1 >= 0) {
                if(arr[cx - 1] != -1) {
                    if(arr[cx - 1] > cc + 1) {
                        arr[cx - 1] = cc + 1
                        q.add(cx - 1 to cc + 1)
                    }
                } else {
                    arr[cx - 1] = cc + 1
                    q.add(cx - 1 to cc + 1)
                }
            }
            if(cx * 2 < MAX) {
                if(arr[cx * 2] != -1) {
                    if(arr[cx * 2] > cc) {
                        arr[cx * 2] = cc
                        q.add(cx * 2 to cc)
                    }
                } else {
                    arr[cx * 2] = cc
                    q.add(cx * 2 to cc)
                }
            }
        }
    }

    private lateinit var check: Array<Boolean>

    fun solution2() {
        System.setIn(FileInputStream(INPUT))
        val br = System.`in`.bufferedReader()
        val (n, k) = br.readLine().split(" ").map { it.toInt() }
        check = Array(MAX) { false }
        val q = PriorityQueue<Pair<Int, Int>> { a, b ->
            a.second - b.second
        }
        q.add(n to 0)
        while(q.isNotEmpty()) {
            val cx = q.peek().first
            val cc = q.peek().second
            println("cx : $cx, cc : $cc")
            q.poll()
            check[cx] = true
            if(cx == k) {
                println(cc)
                return
            }
            if(cx * 2 < MAX && !check[cx * 2]) {
                q.add(cx * 2 to cc)
            }
            if(cx + 1 < MAX && !check[cx + 1]) {
                q.add(cx + 1 to cc + 1)
            }
            if(cx - 1 >= 0 && !check[cx - 1]) {
                q.add(cx - 1 to cc + 1)
            }
        }
    }
}

fun main() {
    HideAndSeek3().solution2()
}